### PR TITLE
Allow multiple element sides on same sideset for elemental boundary auxkernel

### LINKED ISSUE #17424: Less restrictive error for boundary-restricted auxkernel
## Reason

For conjugate heat transfer coupling of nekRS and MOOSE via Cardinal, we use a boundary-restricted auxkernel that computes a CONSTANT MONOMIAL conductive heat flux, `flux`, based on a LINEAR LAGRANGE temperature. This heat flux is then sent to nekRS as a boundary condition between the fluid and solid phases. We would like to run problems where we compute the `flux` on sidesets where elements may have more than one face on this sideset. However, this results in the following error:

> Boundary restricted auxiliary kernel 'flux' has element (id=4023) connected with more than one boundary sides.

Unfortunately, we need this auxkernel to be boundary-restricted because it relies on the normal to compute `k * grad(T) * hat(n)`. But we cannot split the auxkernel into multiple auxiliary variables (as the error message suggests as a fix) because we can’t restrict the flux transfer to our nekRS-wrapped app to a surface - the mesh mirror we construct for nekRS is _only_ a surface (AFAIK, you can only add sideset IDs to sides of volume elements). Even if a splitting solution possible, it’s quite tedious - for flow over 3 rectangular blocks, circumventing this error would need 6 * 3 = 18 separate auxvariables and transfers.

## Design
Add a flag that skips the error check, thereby simply applying the auxkernel to all sides of the element on the sideset based on an evaluation on the side evaluated last. 

## Impact
Allow boundary-restricted elemental auxkernels on multiple sides of an element as a numerical approximation with error that should decrease with mesh refinement. Vastly simpler problem setup and mesh generation for Cardinal.

### PR DESCRIPTION
**Pull request by aprilnovak**
This PR adds a flag that lets you skip the error check that an element doesn't have more than one side on a boundary for boundary-restricted elemental auxkernels. Open to other name ideas for this flag!

Closes #17424

### CODE REVIEW

**Comment by aprilnovak** [comment]
Not sure about that test failure, it seems unrelated?

File: framework/src/auxkernels/AuxKernel.C [hunk 1]
```diff
@@ -54,6 +54,11 @@ AuxKernelTempl<ComputeValueType>::validParams()
                         "displacements are provided in the Mesh block "
                         "the undisplaced mesh will still be used.");
   params.addParamNamesToGroup("use_displaced_mesh", "Advanced");
+  params.addParam<bool>("allow_multiple_element_sides",
+                        false,
+                        "Whether to allow an element with multiple sides "
+                        "on the specified boundary to evaluate based on "
+                        "the value on a single side");
```
Comment (loganharbour) [inline framework/src/auxkernels/AuxKernel.C:61]: The hardest part: definitely naming this thing. I do know that the parameter text should be more detailed and along the lines of:

> Whether or not to check for repeated elements in the case of a boundary restricted, elemental aux variable. Setting this to false will allow the contribution to a single element's elemental value(s) from multiple boundary sides on the same element (example: when the restricted boundary exists on two or more sides of an element, like at a corner of the mesh).

Parameter name ideas:

- `check_boundary_restricted`
- `check_multiple_boundary_restricted`

**Review by loganharbour — CHANGES_REQUESTED** [review]
You need to also add the name of the new parameter to the error, so that it's more clear as to how to disable the error

File: framework/src/auxkernels/AuxKernel.C (hunk 1 above)
Comment (aprilnovak) [reply framework/src/auxkernels/AuxKernel.C]: I like your name ideas better :)

Doc string and error message updated!

**Review by loganharbour — CHANGES_REQUESTED** [review]
Would you mind adding two tests for an input that exhibits the multiple sides case? One with the error disabled and one with? Something minimum with `solve = false` and no kernels/nonlinear variables should work fine I think

**Comment by aprilnovak** [comment]
There was already a test for the default case in `boundary_restricted_error_test`, so just had to add `boundary_unrestricted_error_test`.

File: test/tests/auxkernels/element_aux_boundary/tests [hunk 2]
```diff
@@ -24,4 +24,12 @@
     issues = '#5061'
     design = 'syntax/AuxKernels/index.md'
   [../]
+  [./boundary_unrestricted_error_test]
+    type = RunApp
+    input = 'high_order_boundary_aux.i'
+    cli_args = "AuxKernels/real_property/boundary='0 1' AuxKernels/real_property/check_boundary_restricted=false --color off"
+    requirement = "MOOSE shall allow skipping error if a boundary restricted elemental auxiliary kernel is evaluated on an element with multiple boundary sides"
```
Comment (loganharbour) [inline test/tests/auxkernels/element_aux_boundary/tests:31]:
```suggestion
    requirement = "The system shall allow a boundary restricted elemental auxiliary kernel to be evaluated on an element with multiple boundary sides by setting an input parameter."
```

**Review by loganharbour — CHANGES_REQUESTED** [review]
Just a change for the requirement and good to go!

**Review by loganharbour — DISMISSED** [review]

**Comment by aprilnovak** [comment]
This error stumps me - a test is failing that I'm pretty sure shouldn't be affected by my change - do I need to make the parameter I added restartable or something?

**Comment by loganharbour** [comment]
Ah, yes. This is because storing the solution in this case is non-deterministic and depends on the order of execution for the sides, which for some reason is different with `--recover`. Let met think about this.

**Comment by YaqiWang** [comment]
Just disable the recover test and add the reason why we have to do it (make sure that is indeed what is happening).

**Comment by aprilnovak** [comment]
Well what's strange is that the only test that is failing is one I didn't touch - using the default setting of `check_boundary_restricted = true` (i.e. include the error check). This test _was_ enabled before.

**Comment by YaqiWang** [comment]
Oh, you are right, let me take a closer look.

**Comment by YaqiWang** [comment]
The two tests are using the same input file, could be due to this there is an unwanted interference? We do not need the newly added test run in recover mode anyway.

**Comment by loganharbour** [comment]
I prefer to keep recover testing enabled as much as possible

**Comment by arovinelli** [comment]
any news on this PR?

**Comment by lindsayad** [comment]
Ok I'm going to establish why the old test is failing...

**Comment by lindsayad** [comment]
Haha yes this is definitely an interference issue. Apparently the serialization of tests is equal, in recover mode, to:

1. Execute all part1s
2. Execute all part2s

**Review by lindsayad — DISMISSED** [review]
This looks good to me, however I've got a commit on here so not going to merge (nor am I the assignee)

**Review by loganharbour — APPROVED** [review]
I removed my commit. I don't believe it's necessary.

**Pull request merged**